src/server/v2.0/handler/project.go

```diff
@@ -2,33 +2,211 @@ package handler
+	}
+
+	// accept the "public" property to make replication work well with old versions(<=1.2.0)
+	if req.Public != nil && req.Metadata.Public == "" {
+		req.Metadata.Public = strconv.FormatBool(*req.Public)
+	}
+
+	// populate public metadata as false if it isn't set
+	if req.Metadata.Public == "" {
+		req.Metadata.Public = strconv.FormatBool(false)
+	}
+
+	if err := a.validateProjectReq(ctx, req); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	var ownerID int
+	// set the owner as the system admin when the API being called by replication
+	// it's a solution to workaround the restriction of project creation API:
+	// only normal users can create projects
+	if secCtx.IsSolutionUser() {
+		ownerID = 1
+	} else {
+		ownerName := secCtx.GetUsername()
+		user, err := a.userMgr.GetByName(ctx, ownerName)
+		if err != nil {
+			return a.SendError(ctx, err)
+		}
+		ownerID = user.UserID
+	}
+
+	p := &project.Project{
+		Name:       req.ProjectName,
+		OwnerID:    ownerID,
+		RegistryID: lib.Int64Value(req.RegistryID),
+	}
+	lib.JSONCopy(&p.Metadata, req.Metadata)
+
+	projectID, err := a.projectCtl.Create(ctx, p)
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	if req.StorageLimit != nil {
+		referenceID := quota.ReferenceID(projectID)
+		hardLimits := types.ResourceList{types.ResourceStorage: *req.StorageLimit}
+		if _, err := a.quotaCtl.Create(ctx, quota.ProjectReference, referenceID, hardLimits); err != nil {
+			return a.SendError(ctx, fmt.Errorf("failed to create quota for project: %v", err))
+		}
+	}
+
+	// create a default retention policy for proxy project
+	if req.RegistryID != nil {
+		plc := policy.WithNDaysSinceLastPull(projectID, defaultDaysToRetention)
+		// TODO: move the retention controller to `src/controller/retention` and
+		// change to use the default retention controller in `src/controller/retention`
+		retentionID, err := api.GetRetentionController().CreateRetention(plc)
+		if err != nil {
+			return a.SendError(ctx, err)
+		}
+		md := map[string]string{"retention_id": strconv.FormatInt(retentionID, 10)}
+		if err := a.metadataMgr.Add(ctx, projectID, md); err != nil {
+			return a.SendError(ctx, err)
+		}
+		return nil
+	}
+
+	location := fmt.Sprintf("%s/%d", strings.TrimSuffix(params.HTTPRequest.URL.Path, "/"), projectID)
+	return operation.NewCreateProjectCreated().WithLocation(location)
+}
+
+func (a *projectAPI) DeleteProject(ctx context.Context, params operation.DeleteProjectParams) middleware.Responder {
+	if err := a.RequireProjectAccess(ctx, params.ProjectID, rbac.ActionDelete); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	result, err := a.deletable(ctx, params.ProjectID)
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	if !result.Deletable {
+		return a.SendError(ctx, errors.PreconditionFailedError(errors.New(result.Message)))
+	}
+
+	if err := a.projectCtl.Delete(ctx, params.ProjectID); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	referenceID := quota.ReferenceID(params.ProjectID)
```

Comment:
delete quota firstly?

Reply:
Not need to delete it first because the delete quota does not depend on the info of the project so we can delete it after the project deleted successfully.

The second reason to delete quota after the project deleted successfully is that the deletion of the project may fail. If the deletion of the project failed we can return directly no need to delete the following quota info.